pypowsybl/perunitnetwork.py

```diff
@@ -0,0 +1,615 @@
+        if 'min_p' in keys:
+            to_update['min_p'] *= self.sn
+        self.update_elements(_pypowsybl.ElementType.BATTERY, to_update)
+
+    def update_dangling_lines(self, df: _pd.DataFrame):
+        """ Update dangling lines with a ``Pandas`` data frame.
+
+        Args:
+            df (DataFrame): the ``Pandas`` data frame
+        """
+        keys = df.columns
+        to_update = df.copy()
+        if 'p0' in keys:
+            to_update['p0'] *= self.sn
+        if 'q0' in keys:
+            to_update['q0'] *= self.sn
+        self.update_elements(_pypowsybl.ElementType.DANGLING_LINE, to_update)
+
+    def update_vsc_converter_stations(self, df: _pd.DataFrame):
+        """ Update VSC converter stations with a ``Pandas`` data frame.
+
+        Args:
+            df (DataFrame): the ``Pandas`` data frame
+        """
+        keys = df.columns
+        to_update = df.copy()
+        if 'reactive_power_setpoint' in keys:
+            to_update['reactive_power_setpoint'] *= self.sn
+        to_update = _pd.merge(to_update, self.get_vsc_converter_stations()['voltage_level_id'],
+                              left_index=True, right_index=True)
+        to_update = _pd.merge(to_update, self.get_voltage_levels()['nominal_v'], left_on='voltage_level_id',
+                              right_index=True)
+        if 'voltage_setpoint' in keys:
+            to_update['voltage_setpoint'] *= to_update['nominal_v']
+        to_update = to_update.drop(['nominal_v', 'voltage_level_id'], axis=1)
+        self.update_elements(_pypowsybl.ElementType.VSC_CONVERTER_STATION, to_update)
+
+    def update_static_var_compensators(self, df: _pd.DataFrame):
+        """ Update static var compensators with a ``Pandas`` data frame.
+
+        Args:
+            df (DataFrame): the ``Pandas`` data frame
+        """
+        keys = df.columns
+        to_update = df.copy()
+        if 'reactive_power_setpoint' in keys:
+            to_update['reactive_power_setpoint'] *= self.sn
+        to_update = _pd.merge(to_update, self.get_static_var_compensators()['voltage_level_id'],
+                              left_index=True, right_index=True)
+        to_update = _pd.merge(to_update, self.get_voltage_levels()['nominal_v'], left_on='voltage_level_id',
+                              right_index=True)
+        if 'voltage_setpoint' in keys:
+            to_update['voltage_setpoint'] *= to_update['nominal_v']
+        to_update = to_update.drop(['nominal_v', 'voltage_level_id'], axis=1)
+        self.update_elements(_pypowsybl.ElementType.STATIC_VAR_COMPENSATOR, to_update)
+
+    def update_hvdc_lines(self, df: _pd.DataFrame):
+        """ Update HVDC lines with a ``Pandas`` data frame.
+
+        Args:
+            df (DataFrame): the ``Pandas`` data frame
+        """
+        keys = df.columns
+        to_update = df.copy()
+        if 'active_power_setpoint' in keys:
+            to_update['active_power_setpoint'] *= self.sn
```

Comment:
It seems that the updating of r is missing.

Reply:
there is no r in static var compensator am I wrong ? but it seems that b field is missing and not only for per unitage.

Reply:
Sorry, the comment was for an updated r missing for HVDC lines.